netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +100,256 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
-        } else if (evt == HttpClientUpgradeHandler.UpgradeEvent.UPGRADE_SUCCESSFUL) {
-          upgradeFuture.set(null);
-          ctx.pipeline().remove(this);
+      if (evt instanceof SslHandshakeCompletionEvent) {
+        SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
+        if (handshakeEvent.isSuccess()) {
+          writeBufferedAndRemove(ctx);
+        } else {
+          failBufferedAndClose(ctx);
         }
       }
+      super.userEventTriggered(ctx, evt);
+    }
+  }
+
+  /**
+   * Buffers all writes until the {@link Channel} is active.
+   */
+  private static class BufferUntilChannelActiveHandler extends AbstractBufferingHandler {
+
+    BufferUntilChannelActiveHandler(ChannelHandler... handlers) {
+      super(handlers);
     }
 
     @Override
-    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
-      super.channelInactive(ctx);
-      if (!upgradeFuture.isDone()) {
-        upgradeFuture.setException(new RuntimeException("Channel closed before upgrade complete"));
-      }
+    public void handlerAdded(ChannelHandlerContext ctx) {
+      writeBufferedAndRemove(ctx);
     }
 
     @Override
-    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
-      super.channelUnregistered(ctx);
-      if (!upgradeFuture.isDone()) {
-        upgradeFuture.setException(
-            new RuntimeException("Handler unregistered before upgrade complete"));
-      }
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+      writeBufferedAndRemove(ctx);
+      super.channelActive(ctx);
+    }
+  }
+
+  /**
+   * Buffers all writes until the HTTP to HTTP/2 upgrade is complete.
+   */
+  private static class BufferingHttp2UpgradeHandler extends AbstractBufferingHandler {
+
+    BufferingHttp2UpgradeHandler(ChannelHandler... handlers) {
+      super(handlers);
     }
 
     @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
-      super.exceptionCaught(ctx, cause);
-      if (!upgradeFuture.isDone()) {
-        upgradeFuture.setException(cause);
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+      // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
+      // which causes the upgrade headers to be added
+      DefaultHttpRequest upgradeTrigger =
+          new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
+      ctx.writeAndFlush(upgradeTrigger);
+      super.channelActive(ctx);
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+      if (evt == UpgradeEvent.UPGRADE_SUCCESSFUL) {
+        writeBufferedAndRemove(ctx);
+      } else if (evt == UpgradeEvent.UPGRADE_REJECTED) {
+        failBufferedAndClose(ctx);
+        ctx.fireExceptionCaught(new Throwable("HTTP/2 upgrade rejected"));
```

Comment:
Since we know this handler is at the end of the pipeline, should this be a channel.fireExceptionCaught instead, so that our HTTP/2 handler can deal with it?